File: lib/digital_view.rb
class DigitalView < UIView
  LetterWidth = 24.0
  LetterHeight = 35.0

  DefaultText = '0'
  DefaultColor = :black
  DefaultPad = 3

  attr_accessor :text
  attr_accessor :pad
  attr_accessor :color

  # designated initializer
  def initWithText(text, color: color, pad: pad)
    self.initWithFrame([[0, 0], [0, 0]])

    self.backgroundColor = :gray

    self.text = text
    self.color = color
    self.pad = pad

    self.sizeToFit
    self
  end

  def intrinsicContentSize
    len = @text ? @text.length : 0
    len = @pad if @pad && @pad > len
    CGSize.new(LetterWidth * len, LetterHeight)
  end

  def initWithText(text, color: color)
    self.initWithText(text, color: color, pad: DefaultPad)
  end
  def initWithText(text)
    self.initWithText(text, color: DefaultColor, pad: DefaultPad)
  end
  def init
    self.initWithText(DefaultText, color: DefaultColor, pad: DefaultPad)
  end

  # Options
  #   text => text to output (default: '0')
  #   color => :black, :blue, or a UIColor instance (default: :black)
  #   pad => minimum number of letters to display (default: 3)
  def self.new(opts={})
    self.alloc.initWithText(opts[:text] || DefaultText,
                     color: opts[:color] || DefaultColor,
                       pad: opts[:pad] || DefaultPad)
  end

  ##|
  ##|  IMAGES
  ##|
  def top_image
    @@top_image ||= generate_image(:top)
  end
  def topleft_image
    @@topleft_image ||= generate_image(:topleft)
  end
  def topright_image
    @@topright_image ||= generate_image(:topright)
  end
  def middle_image
    @@middle_image ||= generate_image(:middle)
  end
  def bottomleft_image
    @@bottomleft_image ||= generate_image(:bottomleft)
  end
  def bottomright_image
    @@bottomright_image ||= generate_image(:bottomright)
  end
  def bottom_image
    @@bottom_image ||= generate_image(:bottom)
  end
  def colon
    @@colon_image ||= generate_image(:colon)
  end

  def generate_image(location, color=self.color)
    case color
    when :blue, :black, :gray
      UIImage.imageNamed("digital-#{color.to_s}-#{location.to_s}")
    else
      nil
    end
  end

  ##|
  ##|  SETTERS
  ##|
  def text=(text)
    @text = text.to_s
    setNeedsDisplay
    setNeedsUpdateConstraints
    @text
  end

  def pad=(pad)
    @pad = pad
    setNeedsDisplay
    setNeedsUpdateConstraints
    @pad
  end

  def color=(color)
    @color = color
    @top_image = nil
    @topleft_image = nil
    @topright_image = nil
    @middle_image = nil
    @bottomleft_image = nil
    @bottomright_image = nil
    @bottom_image = nil

    setNeedsDisplay

    @color
  end

  def backgroundColor=(val)
    @backgroundColor = val
    super UIColor.clearColor
    @backgroundColor
  end

  def backgroundColor
    @backgroundColor
  end

  ##|
  ##|  DRAWING
  ##|
  def letter_width
    if self.text.length * LetterWidth < self.frame.width
      LetterWidth
    else
      (self.frame.width / self.text.length).floor
    end
  end

  def letter_height
    (letter_width / LetterWidth * LetterHeight).floor
  end

  def letter_top
    (self.bounds.height - letter_height) / 2
  end

  def drawRect(frame)
    text = self.text
    length = [pad, text.length].max
    total_width = length * self.letter_width
    x = (self.bounds.width - total_width) / 2

    if pad and pad > text.length
      (pad - text.length).times do
        # just draws the background
        x += drawLetter(nil, at:x)
      end
    end

    # if text.length > pad
    #   text = text[(text.length - pad)..text.length]
    # end
    text.each_char do |letter|
      x += drawLetter(letter, at:x)
    end
  end

  def drawLetter(letter, at:x)
    draw_background = 'eight'

    case letter and letter.upcase
    when 'A'
      images = [top_image, topleft_image, topright_image, middle_image, bottomleft_image, bottomright_image]
    when 'B'
      images = [top_image, topleft_image, topright_image, middle_image, bottomleft_image, bottomright_image, bottom_image]
    when 'C'
      images = [top_image, topleft_image, bottomleft_image, bottom_image]
    when 'D'
      images = [top_image, topleft_image, topright_image, bottomleft_image, bottomright_image, bottom_image]
    when 'E'
      images = [top_image, topleft_image, middle_image, bottomleft_image, bottom_image]
    when 'F'
      images = [top_image, topleft_image, middle_image, bottomleft_image]
    when 'G'
      images = [top_image, topleft_image, bottomleft_image, bottomright_image, bottom_image]
    when 'H'
      images = [topleft_image, topright_image, middle_image, bottomleft_image, bottomright_image]
    when 'I'
      images = [topleft_image, bottomleft_image]
    when 'J'
      images = [top_image, topright_image, bottomleft_image, bottomright_image, bottom_image]
    when 'L'
      images = [topleft_image, bottomleft_image, bottom_image]
    when 'O'
      images = [top_image, topleft_image, topright_image, bottomleft_image, bottomright_image, bottom_image]
    when 'P'
      images = [top_image, topleft_image, topright_image, middle_image, bottomleft_image]
    when 'S'
      images = [top_image, topleft_image, middle_image, bottomright_image, bottom_image]
    when 'T'
      images = [middle_image, topright_image, bottomright_image]
    when 'U'
      images = [topleft_image, topright_image, bottomleft_image, bottomright_image, bottom_image]
    when 'Y'
      images = [topleft_image, topright_image, middle_image, bottomright_image]
    when 'Z'
      images = [top_image, topright_image, middle_image, bottomleft_image, bottom_image]
    when '-'
      images = [middle_image]
    when '_'
      images = [bottom_image]
    when '0'
      images = [top_image, topleft_image, topright_image, bottomleft_image, bottomright_image, bottom_image]
    when '1'
      images = [topright_image, bottomright_image]
    when '2'
      images = [top_image, topright_image, middle_image, bottomleft_image, bottom_image]
    when '3'
      images = [top_image, topright_image, middle_image, bottomright_image, bottom_image]
    when '4'
      images = [topleft_image, topright_image, middle_image, bottomright_image]
    when '5'
      images = [top_image, topleft_image, middle_image, bottomright_image, bottom_image]
    when '6'
      images = [top_image, topleft_image, middle_image, bottomleft_image, bottomright_image, bottom_image]
    when '7'
      images = [top_image, topright_image, bottomright_image]
    when '8'
      images = [top_image, topleft_image, topright_image, middle_image, bottomleft_image, bottomright_image, bottom_image]
    when '9'
      images = [top_image, topleft_image, topright_image, middle_image, bottomright_image, bottom_image]
    when ':'
      draw_background = false
      images = [colon]
    else
      images = []
    end

    if draw_background && backgroundColor
      bg_image = generate_image(draw_background, backgroundColor)
      if bg_image
        bg_image.drawInRect([[x, letter_top], [letter_width, letter_height]])
      else
        NSLog "Could not create background image with #{backgroundColor.inspect}"
      end
    end

    images.each {|image|
      if image
        image.drawInRect([[x, letter_top], [letter_width, letter_height]])
      else
        NSLog "Could not create image (?)"
      end
    }

    return self.letter_width
  end

  def sizeThatFits(size)
    return super unless text or pad

    length = pad and (!text or pad > text.length) ? pad : text.length
    [length * LetterWidth, LetterHeight]
  end

end

File: Source/DigitalView.swift
////
///  DigitalView.swift
//

class DigitalView: UIView {
    struct Size {
        static let letterSize = CGSize(width: 24, height: 35)
        static let minTextLength = 3
    }
    enum Color: String {
        case black
        case blue
    }

    var text: String = "0" { didSet {
        guard text != oldValue else { return }
        setNeedsDisplay()
        invalidateIntrinsicContentSize()
    } }
    var minTextLength: Int = Size.minTextLength { didSet {
        guard minTextLength != oldValue else { return }
        setNeedsDisplay()
        invalidateIntrinsicContentSize()
    } }
    var color: Color = .black { didSet {
        guard color != oldValue else { return }
        generateAllImages()
        setNeedsDisplay()
        invalidateIntrinsicContentSize()
    } }

    //|
    //|  IMAGES
    //|
    var topImage: UIImage!
    var topLeftImage: UIImage!
    var topRightImage: UIImage!
    var middleImage: UIImage!
    var bottomLeftImage: UIImage!
    var bottomRightImage: UIImage!
    var bottomImage: UIImage!
    var colon: UIImage!
    var background: UIImage!

    fileprivate func generateAllImages() {
        topImage = generateImage("top")
        topLeftImage = generateImage("topLeft")
        topRightImage = generateImage("topRight")
        middleImage = generateImage("middle")
        bottomLeftImage = generateImage("bottomLeft")
        bottomRightImage = generateImage("bottomRight")
        bottomImage = generateImage("bottom")
        colon = generateImage("colon")
        background = generateImage("eight", color: "gray")
    }

    fileprivate func generateImage(_ location: String, color optColor: String? = nil) -> UIImage? {
        let color = optColor ?? self.color.rawValue
        return UIImage(named: "digital-\(color)-\(location)")
    }

    override required init(frame: CGRect) {
        super.init(frame: frame)

        backgroundColor = .clear
        generateAllImages()
    }

    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override var intrinsicContentSize: CGSize {
        let len = CGFloat(max(minTextLength, text.characters.count))
        return CGSize(width: Size.letterSize.width * len, height: Size.letterSize.height)
    }

    //|
    //|  DRAWING
    //|
    func drawLetterWidth() -> CGFloat {
        if CGFloat(text.characters.count) * Size.letterSize.width < frame.width {
            return Size.letterSize.width
        }
        else {
            return floor(frame.width / CGFloat(text.characters.count))
        }
    }

    func drawLetterHeight() -> CGFloat {
        return floor(drawLetterWidth() / Size.letterSize.width * Size.letterSize.height)
    }

    func drawLetterTop() -> CGFloat {
        return (frame.height - drawLetterHeight()) / 2
    }

    override func draw(_ frame: CGRect) {
        let length = max(minTextLength, text.characters.count)
        let totalWidth = CGFloat(length) * self.drawLetterWidth()
        var x = (frame.width - totalWidth) / 2

        if minTextLength > text.characters.count {
            times(minTextLength - text.characters.count) {
                // just draws the background
                drawBackground(at: x)
                x += drawLetterWidth()
            }
        }

        for letter in text.characters.map({ String($0) }) {
            drawLetter(letter, at: x)
            x += drawLetterWidth()
        }
    }

    func drawBackground(at x: CGFloat) {
        background.draw(in: CGRect(
            x: x,
            y: drawLetterTop(),
            width: drawLetterWidth(),
            height: drawLetterHeight()
            ))
    }

    func drawLetter(_ letter: String, at x: CGFloat) {
        var shouldDrawBackground = true
        let images: [UIImage]
        switch letter {
        case "a":
            images = [topImage]
        case "b":
            images = [topRightImage]
        case "c":
            images = [bottomRightImage]
        case "d":
            images = [bottomImage]
        case "e":
            images = [bottomLeftImage]
        case "f":
            images = [topLeftImage]
        case "g":
            images = [middleImage]
        case "-":
            images = [middleImage]
        case "_":
            images = [bottomImage]
        case "0":
            images = [topImage, topLeftImage, topRightImage, bottomLeftImage, bottomRightImage, bottomImage]
        case "1":
            images = [topRightImage, bottomRightImage]
        case "2":
            images = [topImage, topRightImage, middleImage, bottomLeftImage, bottomImage]
        case "3":
            images = [topImage, topRightImage, middleImage, bottomRightImage, bottomImage]
        case "4":
            images = [topLeftImage, topRightImage, middleImage, bottomRightImage]
        case "5":
            images = [topImage, topLeftImage, middleImage, bottomRightImage, bottomImage]
        case "6":
            images = [topImage, topLeftImage, middleImage, bottomLeftImage, bottomRightImage, bottomImage]
        case "7":
            images = [topImage, topRightImage, bottomRightImage]
        case "8":
            images = [topImage, topLeftImage, topRightImage, middleImage, bottomLeftImage, bottomRightImage, bottomImage]
        case "9":
            images = [topImage, topLeftImage, topRightImage, middleImage, bottomRightImage, bottomImage]
        case ":":
            shouldDrawBackground = false
            images = [colon]
        default:
            images = []
        }

        if shouldDrawBackground {
            drawBackground(at: x)
        }

        let rect = CGRect(
            x: x,
            y: drawLetterTop(),
            width: drawLetterWidth(),
            height: drawLetterHeight()
            )
        for image in images {
            image.draw(in: rect)
        }
    }

    override func sizeThatFits(_ size: CGSize) -> CGSize {
        let length = max(minTextLength, text.characters.count)
        return CGSize(width: CGFloat(length) * Size.letterSize.width, height: Size.letterSize.height)
    }

}
